CosmicReco/src/CosmicTrackFit.cc

```diff
@@ -0,0 +1,654 @@
+	A[1][2] = axes._ZPrime.Y();
+	
+	A[2][0] = axes._XDoublePrime.Z();
+	A[2][1] = axes._YDoublePrime.Z();
+	A[2][2] = axes._ZPrime.Z();
+
+	TMatrixD P(3,1);
+	P[0][0] = Position.X();
+	P[1][0] = Position.Y();
+	P[2][0] = Position.Z();
+        
+	TMatrixD D(3,1);
+	D[0][0] = Direction.X();
+	D[1][0] = Direction.Y();
+	D[2][0] = Direction.Z();
+        
+	TMatrixD PXYZ(A*P);
+	TMatrixD DXYZ(A*D);
+	
+	DXYZ[0][0] = DXYZ[0][0]/ DXYZ[2][0];
+	DXYZ[1][0] = DXYZ[1][0]/ DXYZ[2][0];
+	DXYZ[2][0] = DXYZ[2][0]/ DXYZ[2][0];
+
+	PXYZ[0][0] = PXYZ[0][0] - DXYZ[0][0]*PXYZ[2][0]/ DXYZ[2][0];
+	PXYZ[1][0] = PXYZ[1][0]- DXYZ[1][0]*PXYZ[2][0]/ DXYZ[2][0];
+	PXYZ[2][0] = PXYZ[2][0]- DXYZ[2][0]*PXYZ[2][0]/ DXYZ[2][0];
+       /*
+	//TMatrixD sigmaPos(3,1);
+        TMatrixD Cov(3,3);
+        TMatrixD At(A);
+	At.T();
+        for(unsigned i =0; i< 3; i++){
+        	for(unsigned j =0 ; j <3; j++){
+                    
+	            if(i==0 and j==0) {Cov[i][j] =  cosmictrack->FitParams.Covarience.sigA0; }
+                    if(i==1 and j ==1) { Cov[i][j] = cosmictrack->FitParams.Covarience.sigA1;}
+		    else {Cov[i][j] =0;}
+        	}
+        }
+        TMatrixD CovXYZ(A*Cov*At);
+	for(unsigned i=0; i<3 ; i++){
+	
+		for(unsigned j=0; j<3; j++){
+			cout<<"i= "<<i<<" j= "<<j<<" "<<Cov[i][j]<<"=======> "<<CovXYZ[i][j]<<endl;
+			
+		}
+		
+	}
+       */
+	TMatrixD sigmaPos(3,1);
+	sigmaPos[0][0] = cosmictrack->FitParams.Covarience.sigA0; 
+	sigmaPos[1][0] = cosmictrack->FitParams.Covarience.sigB0;
+	sigmaPos[2][0] = 0;
+	TMatrixD sigmaPosXYZ(A*sigmaPos);
+
+	TMatrixD sigmaDir(3,1);
+	sigmaDir[0][0] = cosmictrack->FitParams.Covarience.sigA1; 
+	sigmaDir[1][0] = cosmictrack->FitParams.Covarience.sigB1;
+ 	sigmaDir[2][0] = 0;
+	TMatrixD sigmaDirXYZ(A*sigmaDir);
+	//TODO  --> check math!!! Currently not using this in fitter!
+	
+
+	XYZVec Pos(PXYZ[0][0], PXYZ[1][0], PXYZ[2][0]);
+	XYZVec Dir(DXYZ[0][0], DXYZ[1][0] , DXYZ[2][0]);
+    
+	if(seed == true){
+		if (det == true){
+			XYZVec PosInDet = ConvertToDetFrame(Pos);
+			TrackEquation XYZTrack(PosInDet, Dir);
+			cosmictrack->SetTrackEquationXYZ(XYZTrack);
+			cosmictrack->sigmaPos.SetXYZ(sigmaPosXYZ[0][0], sigmaPosXYZ[1][0], sigmaPosXYZ[2][0]);
+			cosmictrack->sigmaDir.SetXYZ(sigmaDirXYZ[0][0], sigmaDirXYZ[1][0], sigmaDirXYZ[2][0]);
+			
+		
+		} else {
+			TrackEquation XYZTrack(Pos, Dir);
+			cosmictrack->SetTrackEquationXYZ(XYZTrack);
+			
+		}}
+	else{
+	TrackEquation XYZTrack(Pos, Dir);
+	cosmictrack->SetMinuitTrackEquation(XYZTrack);
+	
+	}
+
+}
+
+/*----------------Fit MC-------------------------------//
+Can do 2 things:
+  1) Get True pos and dir from StepPointMC i.e. independent of my fit routine
+  2) Applky my fit routine to the MC StrawDigiMC i.e. a way to assess accuracy of fitting routine when compared to reco.
+
+//----------------------------------------------------*/
+void CosmicTrackFit::FitMC(CosmicTrackFinderData& trackData, CosmicTrack* cosmictrack, bool Det){	
```

Comment:
they dont, this is a separate function, you mean I should have separate CosmicTrackFitMC.cc

Reply:
Yes.  Reconstruction classes cannot reference MC truth.  It may be you can run your fit unchanged on data provided by MC truth, then all you need is code to stuff the MC truth values into the right data structures.

Reply:
moved